### PR TITLE
`MultihopEstimateInGivenExactAmountOut` and `MultihopEstimateOutGivenExactAmountIn` panic recovering

### PR DESCRIPTION
**Pull request by pysel**
Closes: #4180

## What is the purpose of the change

Panic recovery in `MultihopEstimateInGivenExactAmountOut` and `MultihopEstimateOutGivenExactAmountIn` 

## Brief Changelog

Adds a defer statement that cathes panic and returns empty struct corresponding to function's return type. Similar to 
https://github.com/osmosis-labs/osmosis/blob/main/x/gamm/keeper/swap.go#L39-L44

## Testing and Verifying

Adds unit tests

### CODE REVIEW

File: x/poolmanager/router_test.go [hunk 1]
```diff
@@ -29,6 +29,53 @@ var (
 	concentratedKeeperType    = reflect.TypeOf(&cl.Keeper{})
 )
 
+// prepare pool function
+type preparePool func() (firstEstimatePoolId, secondEstimatePoolId uint64)
+
+// getPoolStrategy returns a function that will setup pools based on the poolType provided
+var getSetupPoolsStrategy = func(suite *KeeperTestSuite, poolType string, poolDefaultSwapFee sdk.Dec) preparePool {
```
Comment (p0mvn) [inline x/poolmanager/router_test.go:36]: Let's define this as a helper method on the suite at the bottom of the file

File: x/poolmanager/router_test.go [hunk 2]
```diff
@@ -29,6 +29,53 @@ var (
...
+	switch poolType {
+	case "stableswap":
```
Comment (p0mvn) [inline x/poolmanager/router_test.go:38]: Please use an enum instead of hardcoded strings

File: x/poolmanager/router_test.go [hunk 3]
```diff
@@ -652,6 +700,64 @@ func (suite *KeeperTestSuite) TestEstimateMultihopSwapExactAmountIn() {
...
+					{
+						PoolId:        4,
+						TokenOutDenom: baz,
+					},
+				},
+				tokenIn:           sdk.NewCoin(foo, sdk.NewInt(100000)),
+				tokenOutMinAmount: sdk.NewInt(1),
+			},
+			expectPass: true,
+			poolType:   "stableswap",
```
Comment (p0mvn) [inline x/poolmanager/router_test.go:730]: Please use an enum here

File: x/poolmanager/router_test.go [hunk 4]
```diff
@@ -662,51 +768,36 @@ func (suite *KeeperTestSuite) TestEstimateMultihopSwapExactAmountIn() {
...
-			firstEstimatePoolId := suite.PrepareBalancerPoolWithPoolParams(balancer.PoolParams{
-				SwapFee: poolDefaultSwapFee,
-				ExitFee: sdk.NewDec(0),
-			})
-			secondEstimatePoolId := suite.PrepareBalancerPoolWithPoolParams(balancer.PoolParams{
-				SwapFee: poolDefaultSwapFee,
-				ExitFee: sdk.NewDec(0),
-			})
+			setupPoolsStrategy := getSetupPoolsStrategy(suite, test.poolType, poolDefaultSwapFee)
+			firstEstimatePoolId, secondEstimatePoolId := setupPoolsStrategy()
```
Comment (p0mvn) [inline x/poolmanager/router_test.go:772]: What is the reason for returning a callback function only to call it on the next line? Can we have only `setupPools` function that returns the ids?

File: x/poolmanager/router_test.go [hunk 5]
```diff
@@ -1030,3 +1118,43 @@ func (suite *KeeperTestSuite) TestSingleSwapExactAmountIn() {
 		})
 	}
 }
+
+// getPoolStrategy returns a function that will setup pools based on the poolType provided
```
Comment (p0mvn) [inline x/poolmanager/router_test.go:1122]:
```suggestion
// setupPools creates pools of desired type anr returns their IDs
```

File: x/poolmanager/router_test.go [hunk 6]
```diff
@@ -1030,3 +1118,43 @@ func (suite *KeeperTestSuite) TestSingleSwapExactAmountIn() {
...
+func (suite *KeeperTestSuite) setupPoolsStrategy(poolType types.PoolType, poolDefaultSwapFee sdk.Dec) (firstEstimatePoolId, secondEstimatePoolId uint64) {
```
Comment (p0mvn) [inline x/poolmanager/router_test.go:1123]: Please rename other instances as well
```suggestion
func (suite *KeeperTestSuite) setupPools(poolType types.PoolType, poolDefaultSwapFee sdk.Dec) (firstEstimatePoolId, secondEstimatePoolId uint64) {
```

**Review by p0mvn — APPROVED** [review]
Nice work

**Comment by p0mvn** [comment]
Merging this since this is blocking v15, and the approach has been discussed before

**Comment by p0mvn** [comment]
Actually, prior to merge, @pysel please add a changelog entry

**Review by ValarDragon — APPROVED** [review]
Nice LGTM! Echo the needs changelog

**Comment by pysel** [comment]
added changelog entry

**Pull request merged**